### PR TITLE
colexec: plan CAST operator for CASE arms when types are mismatched

### LINKED ISSUE #43550: colexec: incorrect type assumptions of CASE operator in some cases
Extracted from https://github.com/cockroachdb/cockroach/issues/43457#issuecomment-569002516

The following interaction
```
root@:26257/defaultdb> create table t(a int2 primary key);
CREATE TABLE

Time: 4.583ms

root@:26257/defaultdb> insert into t values (1);
INSERT 1

Time: 4.326ms

root@:26257/defaultdb> set vectorize_row_count_threshold=0;
SET

Time: 467µs

root@:26257/defaultdb> select case when a = 0 then a else 1:::INT8 end from t;
```
produces an internal error:
```
ERROR: internal error: unexpected error from the vectorized runtime: interface conversion: coldata.column is []int64, not []int16
SQLSTATE: XX000
DETAIL: stack trace:
github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror/error.go:70: func1()
runtime/panic.go:679: gopanic()
runtime/iface.go:255: panicdottypeE()
github.com/cockroachdb/cockroach/pkg/col/coldata/vec.go:194: Int16()
github.com/cockroachdb/cockroach/pkg/col/coldata/vec.eg.go:501: Copy()
github.com/cockroachdb/cockroach/pkg/sql/colexec/allocator.go:118: func1()
github.com/cockroachdb/cockroach/pkg/sql/colexec/allocator.go:90: performOperation()
github.com/cockroachdb/cockroach/pkg/sql/colexec/allocator.go:118: Copy()
github.com/cockroachdb/cockroach/pkg/sql/colexec/case.go:204: Next()
...
```
The problem here is that `CASE` operator has the output column with type `INT2`, but in `ELSE` case we explicitly asked for conversion to `INT8`. The output of `EXPLAIN (TYPES)` shows that `INT2` output type is expected.

Here is `EXPLAIN (VEC)`:
```
  └ Node 1                                
    └ *colexec.caseOp                     
      ├ *colexec.bufferOp                 
      │ └ *colexec.colBatchScan           
      ├ *colexec.projEQInt16Int64ConstOp  
      │ └ *colexec.bufferOp               
      └ *colexec.constInt64Op             
        └ *colexec.bufferOp
```

I'm actually not quite sure what is the correct way to fix this. Possibly we could plan casts to the case's output type in all `WHEN` and `ELSE` arms (when needed). cc @jordanlewis thoughts?

### PR DESCRIPTION
**Pull request by yuzefovich**
CASE operator has predefined output type that all THEN arms as well as
the ELSE arm must respect. However, we also allow for comparable types
to be output. This caused a type conversion error in the vectorized
engine and is now fixed by planning an explicit (although internal) CAST
to the desired output type (when needed).

Fixes: #43550.

Release note (bug fix): Previously, an internal error could occur when
CASE operator operating on distinct although compatible types was
executed via the vectorized engine, and now this is fixed. For example,
a query similar to 'SELECT CASE WHEN false THEN 0:::INT2 ELSE 1:::INT8 END'
previously would error out.

### CODE REVIEW

File: pkg/sql/colexec/execplan.go [hunk 1]
```diff
@@ -1458,10 +1486,23 @@ func planProjectionOperators(
...
+			)
+			if err != nil {
+				return nil, resultIdx, ct, internalMemUsed, err
+			}
+		}
 
 		op := NewCaseOp(NewAllocator(ctx, acc), buffer, caseOps, elseOp, thenIdxs, caseOutputIdx, caseOutputType)
 		internalMemUsed += op.(InternalMemoryOperator).InternalMemoryUsage()
-		return op, caseOutputIdx, ct, internalMemUsed, nil
+		return op, caseOutputIdx, ct, internalMemUsed, err
```
Comment (rohany) [inline pkg/sql/colexec/execplan.go]: nit, but why is this changed?

**Review by rohany — APPROVED** [review]
LGTM, small nit

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43557)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @rohany)

___
*[pkg/sql/colexec/execplan.go, line 1505 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43557#-Lx2fvTTnRvksUh4PNd5-r1-1505:-Lx31zKA-cJShYVdYTjf:b-48khoa) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f1381dd277ff146a75279bf8abccea7a26fc535/pkg/sql/colexec/execplan.go#L1505)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

nit, but why is this changed?
</blockquote></details>

Uhm, no particular reason, I recalled Andrew Werner's recent work on adding a linter about returning errors (although it is not applicable here) and thought that - just to be safe - we should return `err`. I believe it doesn't actually change anything because we check for non-nil errors explicitly. If you think this change doesn't belong here, I'll remove it.

**Comment by rohany** [comment]
I was just wondering if there was a codepath I missed which is why we needed to return err, thats all

**Review by yuzefovich — COMMENTED** [review]
Defensive programming techniques lol :)

TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43557#-:-Lx32mj99GgRWUU9dBMt:b-4yze4c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @rohany)

**Review by yuzefovich — COMMENTED** [review]
Noticed a small nit.

bors r-

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43557#-:-Lx32wLQAchzQmVZfnPo:b-e3fsah)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @rohany)

**Review by yuzefovich — COMMENTED** [review]
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43557#-:-Lx33AmL-JabkN7ByVNk:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @rohany)

**Pull request merged**